bolt-contracts/src/contracts/BoltChallenger.sol

```diff
@@ -98,92 +98,130 @@ contract BoltChallenger is IBoltChallenger {
-            // Refund the excess value, if any
-            payable(msg.sender).transfer(msg.value - CHALLENGE_BOND);
+        if (commitments.length == 0) {
+            revert EmptyCommitments();
+        }
+
+        // Check that the attached bond amount is correct
+        if (msg.value != CHALLENGE_BOND) {
+            revert IncorrectChallengeBond();
         }
 
-        if (commitment.slot > BeaconChainUtils._getCurrentSlot() - BeaconChainUtils.JUSTIFICATION_DELAY_SLOTS) {
+        uint256 targetSlot = commitments[0].slot;
+        if (targetSlot > BeaconChainUtils._getCurrentSlot() - BeaconChainUtils.JUSTIFICATION_DELAY_SLOTS) {
             // We cannot open challenges for slots that are not finalized by Ethereum consensus yet.
             // This is admittedly a bit strict, since 32-slot deep reorgs are very unlikely.
             revert BlockIsNotFinalized();
         }
 
-        // Reconstruct the commitment digest: `keccak( keccak(signed tx) || le_bytes(slot) )`
-        bytes32 commitmentID = _computeCommitmentID(commitment);
+        // Check that all commitments are for the same slot and signed by the same sender
+        // and store the parsed transaction data for each commitment
+        TransactionData[] memory transactionsData = new TransactionData[](commitments.length);
+        (address txSender, address commitmentSigner, TransactionData memory firstTransactionData) =
+            _recoverCommitmentData(commitments[0]);
+
+        transactionsData[0] = firstTransactionData;
+
+        for (uint256 i = 1; i < commitments.length; i++) {
+            (address otherTxSender, address otherCommitmentSigner, TransactionData memory otherTransactionData) =
+                _recoverCommitmentData(commitments[i]);
+
+            transactionsData[i] = otherTransactionData;
+
+            // check that all commitments are for the same slot
+            if (commitments[i].slot != targetSlot) {
+                revert UnexpectedMixedSlots();
+            }
+
+            // check that all commitments are signed by the same sender
+            if (otherTxSender != txSender) {
+                revert UnexpectedMixedSenders();
+            }
+
+            // check that all commitments are signed by the same signer (aka "operator")
+            if (otherCommitmentSigner != commitmentSigner) {
+                revert UnexpectedMixedSigners();
+            }
+
+            // check that the nonces are strictly sequentially increasing in the bundle
+            if (otherTransactionData.nonce != transactionsData[i - 1].nonce + 1) {
+                revert UnexpectedNonceOrder();
+            }
+        }
 
-        // Verify the commitment signature against the digest
-        address commitmentSigner = ECDSA.recover(commitmentID, commitment.signature);
+        // Build the challenge ID: `keccak( keccak(signed tx 1) || keccak(signed tx 2) || ... || le_bytes(slot) )`
+        bytes32 challengeID = _computeChallengeID(commitments);
```

Comment:
this leaves an edge case open:
* proposer commits to 10 transactions (same sender, same slot)
* proposer includes only the first 3, the last 7 are broken commitments
* problem: challengeID should be unique for that sender and slot, but the challenge could be opened with any amount of signed commitments as calldata

A granular solution would be to keep track of all the broken commitments as tx hashes in storage in a mapping of (bytes32 -> bool). Thinking of possible alternatives.

Reply:
I've seen that now the signature is computing by concatenating signatures which is good but still leaves this edge case exposed. I remember the latest proposal on it is to create a challenge _per_ commitment breached, but I don't see this logic in the contract otherwise the function `_computeChallengeID` would return `bytes32[] memory challengeIDs`. Is there any reason we're not doing it?